Add support for session_token

BenKanouse opened the pull request:
If a user is authenticating to S3 using a temporary security token that was obtained through a call to AWS Security Token Service, the S3::Client allows a user to set the `session_token` option. In AWS this option ends up setting the `X-Amz-Security-Token` param in the url.

This commit adds support to the FasterS3Url::Builder that allows a user to set a `session_token` which will intern set the `X-Amz-Security-Token` param in the url.

The specs added confirm it matches the behavior in when a session_token is added to the `Aws::S3::Client` class.

Performance tests:

<details>

<summary>presigned_bench -- This commit</summary>

```
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3   170.000 i/100ms
aws-sdk-s3 with custom headers
                       158.000 i/100ms
 re-used FasterS3Url     2.492k i/100ms
re-used FasterS3Url with cached signing keys
                         3.491k i/100ms
re-used FasterS3URL with custom headers
                         2.047k i/100ms
new FasterS3URL Builder each time
                         2.193k i/100ms
re-used WT::S3Signer    10.272k i/100ms
new WT::S3Signer each time
                         2.850k i/100ms
Calculating -------------------------------------
          aws-sdk-s3      1.748k (± 1.5%) i/s  (572.04 μs/i) -      8.840k in   5.077587s
aws-sdk-s3 with custom headers
                          1.579k (± 1.3%) i/s  (633.47 μs/i) -      7.900k in   5.018456s
 re-used FasterS3Url     24.679k (± 0.8%) i/s   (40.52 μs/i) -    124.600k in   5.054031s
re-used FasterS3Url with cached signing keys
                         42.433k (± 0.7%) i/s   (23.57 μs/i) -    212.951k in   5.023465s
re-used FasterS3URL with custom headers
                         21.973k (± 0.7%) i/s   (45.51 μs/i) -    110.538k in   5.034977s
new FasterS3URL Builder each time
                         23.888k (± 0.6%) i/s   (41.86 μs/i) -    120.615k in   5.051815s
re-used WT::S3Signer    102.459k (± 0.9%) i/s    (9.76 μs/i) -    513.600k in   5.019763s
new WT::S3Signer each time
                         25.473k (± 2.5%) i/s   (39.26 μs/i) -    128.250k in   5.074233s
                   with 95.0% confidence
```
</details>

<details>

<summary>presigned_bench -- v1.1</summary>

```
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3   173.000 i/100ms
aws-sdk-s3 with custom headers
                       157.000 i/100ms
 re-used FasterS3Url     2.656k i/100ms
re-used FasterS3Url with cached signing keys
                         4.581k i/100ms
re-used FasterS3URL with custom headers
                         2.222k i/100ms
new FasterS3URL Builder each time
                         2.552k i/100ms
re-used WT::S3Signer    10.121k i/100ms
new WT::S3Signer each time
                         2.780k i/100ms
Calculating -------------------------------------
          aws-sdk-s3      1.756k (± 1.3%) i/s  (569.38 μs/i) -      8.823k in   5.036679s
aws-sdk-s3 with custom headers
                          1.552k (± 1.7%) i/s  (644.22 μs/i) -      7.850k in   5.080740s
 re-used FasterS3Url     25.827k (± 1.9%) i/s   (38.72 μs/i) -    130.144k in   5.068917s
re-used FasterS3Url with cached signing keys
                         45.096k (± 2.4%) i/s   (22.17 μs/i) -    224.469k in   5.021065s
re-used FasterS3URL with custom headers
                         21.216k (± 1.5%) i/s   (47.13 μs/i) -    106.656k in   5.042634s
new FasterS3URL Builder each time
                         25.391k (± 1.0%) i/s   (39.38 μs/i) -    127.600k in   5.032678s
re-used WT::S3Signer     95.345k (± 1.8%) i/s   (10.49 μs/i) -    475.687k in   5.011579s
new WT::S3Signer each time
                         26.239k (± 2.2%) i/s   (38.11 μs/i) -    130.660k in   5.010333s
                   with 95.0% confidence
```

</details>

BenKanouse committed:
Add support for session_token

If a user is authenticating to S3 using a temporary security token
that was obtained through a call to AWS Security Token Service, the
S3::Client allows a user to set the `session_token` option. In AWS this
option ends up setting the `X-Amz-Security-Token` param in the url.

This commit adds support to the FasterS3Url::Builder that allows a user
to set a `session_token` which will intern set the `X-Amz-Security-Token`
param in the url.

The specs added confirm it matches the behavior in when a session_token
is added to the `Aws::S3::Client` class.

Performance tests:

<details>

<summary>presigned_bench -- This commit</summary>

```
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3   170.000 i/100ms
aws-sdk-s3 with custom headers
                       158.000 i/100ms
 re-used FasterS3Url     2.492k i/100ms
re-used FasterS3Url with cached signing keys
                         3.491k i/100ms
re-used FasterS3URL with custom headers
                         2.047k i/100ms
new FasterS3URL Builder each time
                         2.193k i/100ms
re-used WT::S3Signer    10.272k i/100ms
new WT::S3Signer each time
                         2.850k i/100ms
Calculating -------------------------------------
          aws-sdk-s3      1.748k (± 1.5%) i/s  (572.04 μs/i) -      8.840k in   5.077587s
aws-sdk-s3 with custom headers
                          1.579k (± 1.3%) i/s  (633.47 μs/i) -      7.900k in   5.018456s
 re-used FasterS3Url     24.679k (± 0.8%) i/s   (40.52 μs/i) -    124.600k in   5.054031s
re-used FasterS3Url with cached signing keys
                         42.433k (± 0.7%) i/s   (23.57 μs/i) -    212.951k in   5.023465s
re-used FasterS3URL with custom headers
                         21.973k (± 0.7%) i/s   (45.51 μs/i) -    110.538k in   5.034977s
new FasterS3URL Builder each time
                         23.888k (± 0.6%) i/s   (41.86 μs/i) -    120.615k in   5.051815s
re-used WT::S3Signer    102.459k (± 0.9%) i/s    (9.76 μs/i) -    513.600k in   5.019763s
new WT::S3Signer each time
                         25.473k (± 2.5%) i/s   (39.26 μs/i) -    128.250k in   5.074233s
                   with 95.0% confidence
```
</details>

<details>

<summary>presigned_bench -- v1.1</summary>

```
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-darwin19]
Warming up --------------------------------------
          aws-sdk-s3   173.000 i/100ms
aws-sdk-s3 with custom headers
                       157.000 i/100ms
 re-used FasterS3Url     2.656k i/100ms
re-used FasterS3Url with cached signing keys
                         4.581k i/100ms
re-used FasterS3URL with custom headers
                         2.222k i/100ms
new FasterS3URL Builder each time
                         2.552k i/100ms
re-used WT::S3Signer    10.121k i/100ms
new WT::S3Signer each time
                         2.780k i/100ms
Calculating -------------------------------------
          aws-sdk-s3      1.756k (± 1.3%) i/s  (569.38 μs/i) -      8.823k in   5.036679s
aws-sdk-s3 with custom headers
                          1.552k (± 1.7%) i/s  (644.22 μs/i) -      7.850k in   5.080740s
 re-used FasterS3Url     25.827k (± 1.9%) i/s   (38.72 μs/i) -    130.144k in   5.068917s
re-used FasterS3Url with cached signing keys
                         45.096k (± 2.4%) i/s   (22.17 μs/i) -    224.469k in   5.021065s
re-used FasterS3URL with custom headers
                         21.216k (± 1.5%) i/s   (47.13 μs/i) -    106.656k in   5.042634s
new FasterS3URL Builder each time
                         25.391k (± 1.0%) i/s   (39.38 μs/i) -    127.600k in   5.032678s
re-used WT::S3Signer     95.345k (± 1.8%) i/s   (10.49 μs/i) -    475.687k in   5.011579s
new WT::S3Signer each time
                         26.239k (± 2.2%) i/s   (38.11 μs/i) -    130.660k in   5.010333s
                   with 95.0% confidence
```

</details>

jrochkind commented on lib/faster_s3_url/builder.rb:
```diff
@@ -103,34 +104,24 @@ def presigned_url(key, time: nil, expires_in: DEFAULT_EXPIRES_IN,
-
-      canonical_query_string = canonical_query_string_parts.join("&")
-
+      canonical_query_string = canonical_query_params.collect {|k, v| "#{k}=#{v}" }.join("&")
```
Looks like you removed the `if extra_params.size > 0` check, was that intentional?  I'm not sure if it matters, I guess it was supposed to increase performance in the 0 element case, but it may not actually, is that what you were thinking?

BenKanouse replied:
That is what I was thinking.

I also needed to conditionally add `X-Amz-Security-Token` and the `.compact` call is removing all the nils we don't need for us, so the logic felt simpler to me.

The performance result seemed more or less the same to me.

jrochkind replied:
OK sounds good

BenKanouse commented on lib/faster_s3_url/builder.rb:
```diff
@@ -103,34 +104,24 @@ def presigned_url(key, time: nil, expires_in: DEFAULT_EXPIRES_IN,
+      # These have to be sorted, but sort is case-sensitive, and we have a fixed
+      # list of headers we know might be here... turns out they are already sorted?
+      canonical_query_params = {
+        'X-Amz-Algorithm': ALGORITHM,
```
I noticed that I didn't consistently use double quotes here. It seems like the pattern of the file is to use double quotes for everything. I'll change this line and update my commit.

jrochkind replied:
honestly on small projects like this I don't really worry about that

BenKanouse commented:
I could update the readme [here](https://github.com/jrochkind/faster_s3_url?tab=readme-ov-file#automatic-aws-credentials-lookup) pointing out that a user could fetch the `session_token` from `credentials` if they use this feature.

jrochkind commented:
Looks great, thanks! Sure, if you want to update the README go for it. 

Can you add an entry to CHANGELOG.md in this PR, with a link to this PR, and your name for credit?

BenKanouse commented:
Great!

I will make those changes. I am also going to test out my code in my staging environment to confirm it will work for my use case.

BenKanouse commented:
Hello!

I completed the following:
* Updated the README.md
* Updated the CHANGELOG.md
* Ran this branch in my staging environment and confirmed that the feature worked for me.

Thanks!

jrochkind commented:
Thank you!
